### PR TITLE
improve performance of cleanup BR

### LINKED ISSUE #4073: Improve the performance of cleanup backup files
## Feature Request

**Is your feature request related to a problem? Please describe:**

Currently, when we clean up the backup files in the S3/GCS, the objects are deleted one by one, refer to the code [here](https://github.com/pingcap/tidb-operator/blob/master/cmd/backup-manager/app/clean/clean.go#L40), however, when the data scale is very big, there may be 1 Million files for just one backup, which will take a very long time to clean one backup.
So we should find a better way to speed up the clean-up operation.
**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by KanShiori**
### What problem does this PR solve?

Close #4073
* improve performance of cleanup br
* add e2e test for cleanup br

### Code changes

- [x] Has Go code change
- [x] Has CI related scripts change

### Tests

- [x] Unit test 
- [x] E2E test 
- [x] Manual test 
1. Deploy `Backup` with clean policy 'Delete'.
2. Add files to bucket as backup files.
2. Delete `Backup`.
3. Watch log, and check if all files in storage are deleted.
* AWS S3: 10000 files takes 24 seconds
* NFS: 10000 files takes 17 seconds
* GCS: 10000 files takes 43 seconds
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Improve performance of cleanup `BR`
```

### CODE REVIEW

File: cmd/backup-manager/app/clean/clean.go [hunk 1]
```diff
@@ -43,22 +51,45 @@ func (bo *Options) cleanBRRemoteBackupData(ctx context.Context, backup *v1alpha1
...
-		klog.Infof("Prepare to delete %s for cluster %s", obj.Key, bo)
-		err = s.Delete(context.Background(), obj.Key)
-		if err != nil {
-			return err
+
+		// batch delete objects
+		var s3cli *s3.S3
+		var result *util.BatchDeleteObjectsResult
+		if ok := s.As(&s3cli); ok {
+			klog.Infof("Delete a page of object by S3 batch delete")
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/clean/clean.go]: Please add the backup CR info.

File: cmd/backup-manager/app/clean/clean.go [hunk 2]
```diff
@@ -43,22 +51,45 @@ func (bo *Options) cleanBRRemoteBackupData(ctx context.Context, backup *v1alpha1
...
+				objs[i].Key = fmt.Sprintf("%s/%s", prefix, objs[i].Key)
+			}
+			result = util.BatchDeleteObjectsOfS3(ctx, s3cli, bucket, objs, pageConcurrency)
+		} else {
+			klog.Infof("Delete a page of object concurrently")
+			result = util.BatchDeleteObjectsConcurrently(ctx, s, objs, goroutineConcurrency)
+		}
+
+		if len(result.Deleted) != 0 {
+			klog.Infof("Delete these objects for cluster successfully: %s", strings.Join(result.Deleted, ","))
```
Comment (csuzhangxc) [inline cmd/backup-manager/app/clean/clean.go:83]: too many logs?

File: cmd/backup-manager/app/clean/clean.go [hunk 3]
```diff
@@ -43,22 +51,45 @@ func (bo *Options) cleanBRRemoteBackupData(ctx context.Context, backup *v1alpha1
...
+
+		// batch delete objects
+		var s3cli *s3.S3
+		var result *util.BatchDeleteObjectsResult
+		if ok := s.As(&s3cli); ok {
+			klog.Infof("Delete a page of object by S3 batch delete")
+			bucket := backup.Spec.StorageProvider.S3.Bucket
+			prefix := backup.Spec.StorageProvider.S3.Prefix
+			for i := range objs {
+				objs[i].Key = fmt.Sprintf("%s/%s", prefix, objs[i].Key)
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/clean/clean.go]: prefix may be "", "/", "//", etc. and these cases have been handled in the `cmd/backup-manager/app/util/remote.go`.

File: cmd/backup-manager/app/clean/clean.go [hunk 4]
```diff
@@ -43,22 +51,45 @@ func (bo *Options) cleanBRRemoteBackupData(ctx context.Context, backup *v1alpha1
...
+			}
+			result = util.BatchDeleteObjectsOfS3(ctx, s3cli, bucket, objs, pageConcurrency)
+		} else {
+			klog.Infof("Delete a page of object concurrently")
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/clean/clean.go]: Please also test GCS/NFS and update the result in the PR description.

File: cmd/backup-manager/app/clean/clean.go (hunk 1 above)
Comment (DanielZhangQD) [reply cmd/backup-manager/app/clean/clean.go]: Please also check the other logs.

File: cmd/backup-manager/app/clean/clean.go [hunk 5]
```diff
@@ -43,22 +51,45 @@ func (bo *Options) cleanBRRemoteBackupData(ctx context.Context, backup *v1alpha1
...
 		}
-		klog.Infof("Prepare to delete %s for cluster %s", obj.Key, bo)
-		err = s.Delete(context.Background(), obj.Key)
-		if err != nil {
-			return err
+
+		// batch delete objects
+		var s3cli *s3.S3
+		var result *util.BatchDeleteObjectsResult
+		if ok := s.As(&s3cli); ok {
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/clean/clean.go]: This check should be done outside of the `for` loop?

File: cmd/backup-manager/app/clean/clean.go [hunk 6]
```diff
@@ -18,14 +18,22 @@ import (
...
+	"github.com/aws/aws-sdk-go/service/s3"
 	"github.com/pingcap/tidb-operator/cmd/backup-manager/app/constants"
 	"github.com/pingcap/tidb-operator/cmd/backup-manager/app/util"
 	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
 )
 
+const (
+	pageSize             = 10000
+	pageConcurrency      = 10
+	goroutineConcurrency = 100
```
[the remark this answers is not in the document]
Comment (KanShiori) [reply cmd/backup-manager/app/clean/clean.go:33]: It means we get 10000 objects, then delete them.

File: tests/e2e/br/br.go [hunk 7]
```diff
@@ -215,6 +215,60 @@ var _ = ginkgo.Describe("Backup and Restore", func() {
...
+			err = createRBAC(f)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Create backup with clean policy Delete")
+			backup, err := createBackupAndWaitForComplete(f, backupName, backupClusterName, typ, func(backup *v1alpha1.Backup) {
+				backup.Spec.CleanPolicy = v1alpha1.CleanPolicyTypeDelete
+			})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Delete backup")
```
[the remark this answers is not in the document]
Comment (KanShiori) [reply tests/e2e/br/br.go:261]: function `createBackupAndWaitForComplete` wait for the backup to succeed ?

File: cmd/backup-manager/app/clean/clean.go (hunk 2 above)
Comment (KanShiori) [reply cmd/backup-manager/app/clean/clean.go:69]: I think log is less than before, or remove logs about deleted objects ?

File: cmd/backup-manager/app/clean/clean.go (hunk 3 above)
Comment (KanShiori) [reply cmd/backup-manager/app/clean/clean.go]: We use `blob.PrefixBucket` to `List`, and key in result of `List`  don't contain prefix.
But `s3.S3` use complete key to delete.

File: cmd/backup-manager/app/util/remote.go [hunk 8]
```diff
@@ -64,34 +70,219 @@ type localConfig struct {
...
+
+	return gcsClient, true
+}
+
+// GetBucket return bucket name
+//
+// If provider is S3/GCS, return bucket. Otherwise return empty string
+func (b *StorageBackend) GetBucket() string {
+	if b.provider.S3 != nil {
+		return b.provider.S3.Bucket
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/util/remote.go:145]: Does https://github.com/pingcap/tidb-operator/pull/4149/files affect this?

Comment (DanielZhangQD) [reply cmd/backup-manager/app/util/remote.go:145]: We need to add an e2e for #4149

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 9]
```diff
@@ -1359,6 +1359,24 @@ const (
...
 )
 
+// +k8s:openapi-gen=true
+// CleanOption defines the configuration for cleanup backup
+type CleanOption struct {
+	// PageSize represents the number of objects to clean at a time.
+	// default is 10000
+	PageSize uint64 `json:"pageSize,omitempty"`
+
+	// DisableBatchConcurrency disable batch deletions.
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1369]:
```suggestion
	// DisableBatchConcurrency disables the batch deletions and the cleanup will be done by goroutines. 
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 10]
```diff
@@ -1359,6 +1359,24 @@ const (
...
+	DisableBatchConcurrency bool `json:"disableBatchConcurrency,omitempty"`
+	// BatchConcurrency represents the number of batch deletions in parallel.
+	// It is used when storage provider support batch delete api.
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1372]:
```suggestion
	// It is used when the storage provider supports the batch delete API, currently, S3 only.
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 11]
```diff
@@ -1359,6 +1359,24 @@ const (
...
+	// default is 10
+	BatchConcurrency uint32 `json:"batchConcurrency,omitempty"`
+	// BatchConcurrency represents the number of goroutine that used to delete objects
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1375]:
```suggestion
	// RoutineConcurrency represents the number of goroutines that used to delete objects
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 12]
```diff
@@ -1424,6 +1442,8 @@ type BackupSpec struct {
 	ServiceAccount string `json:"serviceAccount,omitempty"`
 	// CleanPolicy denotes whether to clean backup data when the object is deleted from the cluster, if not set, the backup data will be retained
 	CleanPolicy CleanPolicyType `json:"cleanPolicy,omitempty"`
+	// CleanOption control the behavior of clean.
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1445]:
```suggestion
	// CleanOption controls the behavior of clean.
```

File: tests/e2e/br/br.go [hunk 13]
```diff
@@ -215,6 +215,60 @@ var _ = ginkgo.Describe("Backup and Restore", func() {
...
+			backupDSN := getDefaultDSN(backupHost, dbName)
+			err = blockwriter.NewDefault().Write(context.Background(), backupDSN)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Create RBAC for backup and restore")
+			err = createRBAC(f)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Create backup with clean policy Delete")
+			backup, err := createBackupAndWaitForComplete(f, backupName, backupClusterName, typ, func(backup *v1alpha1.Backup) {
```
Comment (DanielZhangQD) [inline tests/e2e/br/br.go:256]: Suggest including the cleanOptions in the case.

Comment (KanShiori) [reply tests/e2e/br/br.go:256]: OK, I'll add that later

Comment (DanielZhangQD) [reply tests/e2e/br/br.go:256]: We have default values for the clean options, so it should be OK not to add them explicitly.

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 14]
```diff
@@ -1359,6 +1359,32 @@ const (
 	CleanPolicyTypeDelete CleanPolicyType = "Delete"
 )
 
+// BatchDeleteOption controls delete object in batches for cleanup backup
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1362]:
```suggestion
// BatchDeleteOption controls the options to delete the objects in batches during the cleanup of backups
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 15]
```diff
@@ -1359,6 +1359,32 @@ const (
...
+//
+// +k8s:openapi-gen=true
+type BatchDeleteOption struct {
+	// DisableBatchConcurrency disables the batch deletions and the cleanup will be done by goroutines.
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1366]:
```suggestion
	// DisableBatchConcurrency disables the batch deletions with S3 API and the deletion will be done by goroutines.
```

File: tests/e2e/br/br.go [hunk 16]
```diff
@@ -216,6 +225,81 @@ var _ = ginkgo.Describe("Backup and Restore", func() {
...
+		expectedPrefix := ""
+
+		ns := f.Namespace.Name
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		tcase := newTestCase(utilimage.TiDBLatest, utilimage.TiDBLatest, typeBR)
+		tcase.configureBackup = func(backup *v1alpha1.Backup) {
+			backup.Spec.StorageProvider.S3.Bucket = path.Join(backup.Spec.StorageProvider.S3.Bucket, middlePath) // bucket add suffix
+			expectedPrefix = path.Join(middlePath, backup.Spec.StorageProvider.S3.Prefix)
```
Comment (DanielZhangQD) [inline tests/e2e/br/br.go:240]: Is it better to move this to the postBackup block?

Comment (KanShiori) [reply tests/e2e/br/br.go:240]: the function postBackup exec after creating Backup, so we should configure Backup in the function configureBackup

Comment (DanielZhangQD) [reply tests/e2e/br/br.go:240]: I mean the variable `expectedPrefix `.

Comment (KanShiori) [reply tests/e2e/br/br.go:240]: oh, yes 😄

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by july2993 — APPROVED** [review]

**Pull request merged**